manager/state/raft/transport/transport.go

```diff
@@ -0,0 +1,250 @@
+	t.mu.Unlock()
+	if !ok {
+		return ErrIsNotFound
+	}
+	if addr == p.address() {
+		return nil
+	}
+	cc, err := t.dial(addr)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create conn for %x with addr %s", id, addr)
+	}
+	ctx, cancel := context.WithTimeout(ctx, t.config.SendTimeout)
+	defer cancel()
+	resp, err := api.NewHealthClient(cc).Check(ctx, &api.HealthCheckRequest{Service: "Raft"})
+	if err != nil {
+		cc.Close()
+		return errors.Wrap(err, "failed to check health")
+	}
+	if resp.Status != api.HealthCheckResponse_SERVING {
+		cc.Close()
+		return fmt.Errorf("health check returned status %s", resp.Status)
+	}
+	p.update(cc, addr)
+	return nil
+}
+
+func (t *Transport) GetPeerAddr(id uint64) (string, error) {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	p, ok := t.peers[id]
+	if !ok {
+		return "", ErrIsNotFound
+	}
+	return p.address(), nil
+}
+
+func (t *Transport) sendTemporary(ctx context.Context, m raftpb.Message) error {
+	longestActive, err := t.longestActive()
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithTimeout(ctx, 2*t.config.SendTimeout)
+	defer cancel()
+	addr, err := longestActive.resolveAddr(ctx, m.To)
+	if err != nil {
+		return err
+	}
+	p, err := newPeer(m.To, addr, t)
+	if err != nil {
+		return errors.Wrap(err, "creating temporary peer")
+	}
+	defer p.stop()
+	if err := p.sendProcessMessage(ctx, m); err != nil {
+		return errors.Wrap(err, "processing message to temporary peer")
+	}
+	return nil
+}
+
+// Send sends slice of raft messages to remote peers. It returns map of errors,
+// where key is message and value is error.
+// This method doesn't block as it just notifies transport that it should send
+// message. It doesn't return errors from real over-the-network code.
+func (t *Transport) Send(ctx context.Context, ms []raftpb.Message) error {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	if t.stopped {
+		return errors.New("transport stopped")
+	}
+	var errs []error
+	for _, m := range ms {
+		if t.config.TransportRaft.IsIDRemoved(m.To) {
+			errs = append(errs, errors.Errorf("refusing to send message %s to removed member %x", m.Type, m.To))
+			continue
+		}
+		p, ok := t.peers[m.To]
+		if ok {
+			if err := p.send(ctx, m); err != nil {
+				errs = append(errs, errors.Wrapf(err, "failed to send message %x to %x", m.Type, m.To))
+			}
+			continue
+		}
+		go func() {
+			// If we are being asked to send to a member that's not in
+			// our member list, that could indicate that the current leader
+			// was added while we were offline. Try to resolve its address.
+			log.G(ctx).Warningf("sending message %s to an unrecognized member ID %x", m.Type, m.To)
+			if err := t.sendTemporary(ctx, m); err != nil {
+				log.G(ctx).Warnf("ignored message %s to unknown peer %x because of: %v", m.Type, m.To, err)
+			}
+		}()
+	}
+	if len(errs) != 0 {
+		var errStr string
+		for _, e := range errs {
+			errStr += "\n" + e.Error()
+		}
+		return errors.Errorf("errors occured during Send: %s", errStr)
```

Comment:
I assume these eventually end up in the log. I think it's a lot cleaner to just log each error separately. Logging something containing newlines is pretty bad.